### DISCUSSION TITLE
guide/document/

### DISCUSSION THREAD

**Discussion by giscus** [pengzhanbo/vuepress-theme-plume#98] [docs-comment]
# guide/document/

概述 在本主题满足了 Blog的基本功能后，期望能够 以 note 或者 book 的形式聚合文章，形式上类似于 vuepress 默认主题。 同时也减少配置的复杂度。 它能够让你以更加友好的方式，组织管理你的文档，或者 知识笔记。 目录 文档/知识笔记 默认存放在 {sourceDir}/notes 目录下。 示例： 其中，typescript 和 ...

https://theme-plume.vuejs.press/guide/document/

**Comment by Mars13333** [comment]
按照当前页配置notes，没有生成文档/知识笔记。请问有什么需要注意的地方么？

**Reply by pengzhanbo** [reply]
请提供更多的信息，我无法判断你遇到的问题是什么

**Reply by Mars13333** [reply]
我直接提供github连接合适么。。。不知道说，就是按照快速开始，配置notes，导航位置使用没有笔记，就是那个TypeScript和rust目录例子。

**Reply by Mars13333** [reply]
https://github.com/Mars13333/mars-vuepress.git

**Reply by pengzhanbo** [reply]
根据你的仓库的配置信息，我大概理解了你的意思。其实 notes 是有正常工作的。问题的根源在于，你可能是在 配置 notes 之前就已经创建好 rust、typescript 目录以及文件并启动了 vuepress 服务，这导致 主题未读取到 notes 配置就 生成了 frontmatter, 使得 frontmatter  中的 permalink 已经生成完毕了，然后你再写入了 notes 配置，重启 vuepress 服务时，由于 permalink 已经存在，主题也就不再重新生成，这导致了 permalink  没有与 notes 配置产生关联。

我在这个文档的末尾有 加 提示说明需要先配置 notes 再创建文件，因为这经常会被忽略导致 此类问题。
修复此问题很简单，删除 markdown 文件 frontmatter 中 permalink 字段，让主题重新生成即可。

需要注意的是，notes 创建后，链接并不会自动注入到 导航栏中，这需要你自行将 note 添加到 [navbar](https://theme-plume.vuejs.press/config/navigation/) 。

**Reply by Mars13333** [reply]
十分感谢。

**Reply by zhengchen-zzZ** [reply]
大大，貌似文件夹名称是中文的，notes无法生成侧边栏。
比如这样的：
const ProjectNote = defineNoteConfig({
    text: '项目开发',
    dir: '项目开发/',              // 这项配置中文貌似没有生成侧边栏
    link: 'notes/Project/', 
    sidebar: 'auto'
})
而
const ProjectNote = defineNoteConfig({
    text: '项目开发',
    dir: 'Project/',               //配置了可以生成侧边栏
    link: 'notes/Project/',
    sidebar: 'auto'
})
不知道是否是一个bug

**Reply by pengzhanbo** [reply]
我在本地使用中文创建，包括多层嵌套，均能生成侧边栏，未能重现问题。你是否可以提供一个重现包给我进行检查。
https://v2.vuepress.vuejs.org/new 可以使用这个进行在线编辑重现包。

**Reply by zhengchen-zzZ** [reply]
已复现，链接为https://stackblitz.com/edit/vuepress-q4qop5?file=docs%2F.vuepress%2Fplume.config.ts，
可能是我的使用方式有错误，作者考虑过出一个博客来讲解一下你的使用案例吗，对这个notes用的比较迷

**Reply by pengzhanbo** [reply]
中文编码问题。不是 `dir` 字段的问题，是 `link` 字段中包含了中文。我在处理 sidebar 时未考虑 链接中包含 中文 的情况，导致了在对 链接进行 解析时，未匹配到对应的 sidebar 数据。实际上是有生成的。

关于 `notes`，我一开始的想法就是将 `sidebar` 做更语义化的配置，同时映射到文件夹结构上，因为原本 vuepress 的默认主题的 `sidebar` 配置对新手来说并不友好，实际上 `notes` 应该是 https://pengzhanbo.cn/memorandum/ 类似这种，相当于一个 子文档，有可选的单独主页，以及一系列文章的聚合。

主题依然保留了 [`sidebar`](https://theme-plume.vuejs.press/config/basic/#sidebar) 配置项，满足不同的使用习惯。

**Reply by zhengchen-zzZ** [reply]
感谢作者答复，确实vuepress官网那个sidebar配的我很难受，还是作者的notes这个比较适合我

**Comment by tianhairui** [comment]
作者好，我在环境中按照该文档中的文件结构及侧标栏做了对应的配置，侧标栏配置为自动生成，`config.js` 中相关配置如下：
```shell
......
    notes: {
      link: '/',
      dir: '/notes/',
      notes: [
        {
          dir: 'typescript',
          link: '/typescript/',
          sidebar: 'auto'
        },
        {
          dir: 'ruby',
          link: '/ruby/',
          sidebar: [
            {text: '简介', items: ['intro.md']}
          ]
        }
      ], 
    }
......
// 之后，我在 navbar 导航栏中配置如下，起名叫 Book
{ text: 'Book', link: '/article/cyc1oseq/', icon: 'hugeicons:note-01'/*'material-symbols:article-outline'*/ }
```
这里我出现的问题是：
在`navbar`的`link`中，如果指定具体md文档中的的`permalink`，那么点击导航 Book 跳转至对应的文档，也没有侧边栏生成；如果指定具体的notes路径，点击导航 Book 就会出错，折腾了半个月，文档梳理了多遍，还是未能实现，请问我忽略了什么地方？

**Reply by pengzhanbo** [reply]
首先需要明确的是，notes 的配置声明应该早于 其关联的目录和文件的创建，如果目录文件先准备好而 note 配置未声明，启动开发服务时，由于读取不到对应的配置，主题会认为它们应用默认的生成规则，即博客类型文章的规则，此时的 permalink 被错误的生成为 `/arcitle/xxx/`；如果先准备好配置再启动开发服务和创建相关文件，则生成的 permalink 应该是 `/typescript/xxx/` 。那么在 navbar 中使用的就是 `/typescript/xxx/` 或者文件路径。
  如果 permalink 已经被错误的生成了，那么就直接把文章中的 permalink 删除掉，重启 开发服务，由主题重新根据配置生成即可。
   sidebar 侧边栏的数据匹配是根据 链接前缀 的，即 note 中声明的 `link: '/typescript/'` 就是匹配的链接前缀，而相关联的 markdown 文章，其 permalink 也应该是 `'/typescript/xxxx/'` 才能匹配到。

**Reply by tianhairui** [reply]
finally~(❁´◡`❁)😘感谢作者，终于搞好了，超赞！

**Reply by sunshang-hl** [reply]
你的项目开源了么？我想去学习学习，侧边栏我折腾了很久，还没有搞好。

**Reply by pengzhanbo** [reply]
@sunshang-hl 我建议你详细描述下你在配置过程中遇到了哪些问题。我不确定到底是我的文档描述有问题，还是侧边栏配置的学习门槛依然过高。

**Reply by sunshang-hl** [reply]
在开始配置前我以为很简单的,我按照教程操作的,但是侧边栏永远没办法正常显示,在项目demo示例下,我添加新的MD文件,按道理它会和foo一起正常显示对吧.但是并没有.能否建立一个用户交流群.让大家更加便捷的交流.

**Reply by sunshang-hl** [reply]
或者扩展模板项目,增加知识文档笔记侧边栏的相关配置,让用户开箱即用,尽可能完善所有的配置.用户只需要关注写作即可.非常的喜欢这款主题,因此还在继续尝试,解决我遇到的问题.

**Reply by sunshang-hl** [reply]
还有,目前访问这个网站,打开新页面,经常需要加载数十秒,是否对网站的性能进行优化呢?

**Reply by pengzhanbo** [reply]
@sunshang-hl 交流群已建立， discussions  公告有群号。
模板中的 docs/.vuepress/ 下就有 `notes.ts` 配置的 demo。
主题的文档站点是部署在 netlify 上的，服务器在海外，国内网络访问必然不会很快，一般使用下科学，访问速度还是可以的。

**Reply by sunshang-hl** [reply]
已申请入群,建议将群号放在文档首页或其他显眼的位置,便于用户入群交流. 是否可尝试将网站cdn部署在cloudflare,优化访问体验,而且没有费用支出.一点建议.

**Comment by physnya** [comment]
您好，我发现文档中似乎有一个很小的笔误，在“编写 notes 配置 → 笔记较多时怎么配置”中，```.vuepress/config.ts``` 中的第一行增加的代码应该是：
```
import { notes } from './notes/index.ts'
```
不知道我说得对不对，烦请您看一下了。

**Reply by pengzhanbo** [reply]
感谢指正。

**Comment by 00D2** [comment]
请问：defineNoteConfig, defineNotesConfig是不是各自只能有一层，不能进行嵌套。
比如NotesConfig下面再次引入NotesConfig。

**Comment by okkeman** [comment]
每个note下必须添加README.md文件，示例都没添加，新手很难注意到这个问题。

**Reply by pengzhanbo** [reply]
并没有这个要求。

**Reply by okkeman** [reply]
大神，我又新建了多个note，实测结果很奇怪
1. README.md 文件必须添加，不添加就会报404错误
2. 侧边栏点击的时候有时候会自动隐藏

**Reply by pengzhanbo** [reply]
> README.md 文件必须添加，不添加就会报404错误

这说明你的 链接应该指向 note 下存在的 markdown 文件的链接，而不是一个不存在的链接。README.md 只是一个可选的 note 首页，是否需要取决自身的需求。

> 侧边栏点击的时候有时候会自动隐藏

检查你的 markdown 中的 permalink 是否正确，它的前缀应该与 node 配置下的 link 选项相同。侧边栏是根据链接前缀和 permalink 进行匹配的。

**Comment by okkeman** [comment]
您确定 sidbarItem的link开头前要加'/'么？为什么实际运行结果导航到了根目录，去掉这个'/'才能导航到笔记。另外还有个问题：为什么点击item后，侧边栏会消失呢？
const typescript = defineNoteConfig({
  dir: 'typescript',
  link: '/typescript/',
  sidebar: [
    { text: '介绍', link: '/guide/intro' },
    { text: '快速上手', link: '/guide/getting-start' },
  // ...
  ]
})

// ... other code

**Reply by pengzhanbo** [reply]
文档中没有任何一行内容说必须要 / 开头。

**Reply by ykshang** [reply]
他压根没看明白 开头保持统一风格不带`/`， 配置前缀的时候末尾带`/` 就行了。或者保持统一风格使用根路径

**Comment by ykshang** [comment]
反馈一个建议 和一个 bug：
- 建议：能够在配置`"sidebar": "auto"` 的情况下，使用前缀序号`1.`  / `2.` 标记顺序的时候，增加一个额外的选项，希望提取文章名字的时候能够选择是够保留这个序号。
- BUG 另外一个问题，文章名字包含某些特殊字符时，会导致项目 `exit 0`;  如该附件截图：
![微信截图_20250505122303](https://github.com/user-attachments/assets/3bdaff99-54d2-4123-95a4-83c23e3253d7)

**Reply by ykshang** [reply]
补充一下是 `@` 字符作为文件名第一个字符时，导致崩溃。

**Reply by pengzhanbo** [reply]
yaml 字符编码下个版本解决。
关于裁剪序号做成可选的配置，这方面我还需要考虑下是否有必要，因为我在设计这块的时候，考虑是的从 命名约定 的方向做的，序号就仅作为排序时候使用。而且在我看来，在侧边栏中显示序号，有些 “多此一举”。

**Reply by ykshang** [reply]
嗯嗯，主要是，我目前在梳理一些面试题，有些名字超长，导致换行。加个序号显得更整齐。

**Reply by ykshang** [reply]
我是觉得 开放一点，让用户自己决定要不要启用这个配置

**Comment by yabro-666** [comment]
我定义如下文件夹
```
notes  
    -  学习笔记     
          - assets         
                 - a.png          
                 - b.png      
          -  测试文章.md
```
在页面上测试文章引用了assets中的图片，但是并未显示图片。其中md中图片的引用如下![](./assets/a.png)，请问一下作者大大如何解决这种问题

**Reply by yabro-666** [reply]
已解决

**Comment by Niceies** [comment]
是否有左侧导航栏版本选项的功能？

**Reply by pengzhanbo** [reply]
可以通过 布局插槽 + 自定义组件 的方式自行添加相关功能

**Comment by RowanWang6** [comment]
请问根据单个markdown文件的标题级别生成侧边栏，除了手动配置link为md锚点，还有快速实现方法吗
